source/neuropod/bindings/java/src/main/native/com_uber_neuropod_NeuropodTensor.cc

```diff
@@ -109,3 +112,90 @@ JNIEXPORT jlong JNICALL Java_com_uber_neuropod_NeuropodTensor_nativeGetNumberOfE
     }
     return 0;
 }
+
+JNIEXPORT jobject JNICALL Java_com_uber_neuropod_NeuropodTensor_nativeToStringList(JNIEnv *env,
+                                                                                   jclass,
+                                                                                   jlong nativeHandle)
+{
+    try
+    {
+        auto neuropodTensor =
+            (*reinterpret_cast<std::shared_ptr<neuropod::NeuropodValue> *>(nativeHandle))->as_tensor();
+        auto tensorType = neuropodTensor->get_tensor_type();
+        if (tensorType != neuropod::STRING_TENSOR)
+        {
+            throw std::runtime_error("unexpected tensor type, should be STRING_TENSOR but found:" +
+                                     tensorTypeToString(tensorType));
+        }
+
+        auto    size = neuropodTensor->get_num_elements();
+        jobject ret  = env->NewObject(java_util_ArrayList, java_util_ArrayList_, size);
+        if (!ret)
+        {
+            throw std::runtime_error("out of memory: cannot create ArrayList");
+        }
+
+        auto typedTensor = neuropodTensor->as_typed_tensor<std::string>();
+
+        std::function<void(string_accessor_type *)> mapFunc = [env, ret](string_accessor_type *elem) {
+            // StringProxy supports conversion to std:string.
+            std::string tmpStr(*elem);
+            jstring     convertedElem = env->NewStringUTF(tmpStr.c_str());
+            env->CallBooleanMethod(ret, java_util_ArrayList_add, convertedElem);
+            env->DeleteLocalRef(convertedElem);
+        };
+
+        auto dims = typedTensor->get_dims();
+        switch (dims.size())
+        {
+        case 1:
+            mapStringTensor(typedTensor->accessor<1>(), mapFunc, dims);
+            break;
+        case 2:
+            mapStringTensor(typedTensor->accessor<2>(), mapFunc, dims);
+            break;
+        case 3:
+            mapStringTensor(typedTensor->accessor<3>(), mapFunc, dims);
+            break;
+        case 4:
+            mapStringTensor(typedTensor->accessor<4>(), mapFunc, dims);
+            break;
+        default:
+            // Here copy data twice
+            const auto &elementList = typedTensor->get_data_as_vector();
+            for (const auto &elem : elementList)
+            {
+                jstring convertedElem = env->NewStringUTF(elem.c_str());
+                env->CallBooleanMethod(ret, java_util_ArrayList_add, convertedElem);
+                env->DeleteLocalRef(convertedElem);
+            }
+        }
```

Comment:
Can we improve this? Maybe we should just add this as a friend function of `TypedNeuropodTensor<std::string>`. That way, it could have individual element access without a lot of the messiness here.

Also, maybe it's worth benchmarking just always using `get_data_as_vector`.  Each elementwise access to a string tensor goes through virtual calls which may be slow. For larger tensors, it's possible/likely that the branch predictor will do a good job and minimize the impact, but still worth benchmarking.

Reply:
I considered a little but different approach. I added to TypedNeuropodTensor<std::string>:

get_data_at(index)
set_data_at(index, value)

that allows to avoid dealing with intermediate std::vector.

I also removed mapStringTensor+accessor because having get_set_at looks like optimal enough. 

Please look and advise